bgpd/bgp_labelpool.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * BGP Label Pool - Manage label chunk allocations from zebra asynchronously
+ *
+ * Copyright (C) 2018 LabN Consulting, L.L.C.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef _FRR_BGP_LABELPOOL_H
+#define _FRR_BGP_LABELPOOL_H
+
+#include <zebra.h>
+
+#include "fifo.h"
+#include "mpls.h"
+
+/*
+ * Types used in lp_get for debug tracking; add more as needed
+ */
+#define LP_TYPE_VRF	0x00000001
+
+/*
+ * label control block
+ */
+struct lp_lcb {
+	mpls_label_t	label;		/* MPLS_LABEL_NONE = not allocated */
+	int		type;
+	void		*labelid;	/* unique ID */
+	/*
+	 * callback for label allocation and loss
+	 *
+	 * allocated: false = lost
+	 */
+	int		(*cbfunc)(mpls_label_t label, void *lblid, bool alloc);
+};
+
+struct lp_chunk {
+	uint32_t	first;
+	uint32_t	last;
+};
+
+/* XXX same first elements as "struct fifo" */
+struct lp_fifo {
+	struct lp_fifo	*next;
+	struct lp_fifo	*prev;
+
+	uint32_t	count;
+	struct lp_lcb	lcb;
+};
+
+struct labelpool {
+	struct skiplist		*ledger;	/* all requests */
+	struct skiplist		*inuse;		/* individual labels */
+	struct list		*chunks;	/* granted by zebra */
+	struct lp_fifo		requests;	/* blocked on zebra */
+	struct work_queue	*callback_q;
+	uint32_t		pending_count;	/* requested from zebra */
+};
+
+#define LABEL_FIFO_ADD(F, N)						\
```

Comment:
Do we really need a fifo here? If yes, wouldn't it be possible to improve `lib/fifo.[ch]` so we don't need to replicate this code every time this data structure is used?

Reply:
It seemed like the closest match to what I needed to process the requests in order, although I admit I have not studied everything in lib. While I agree that lib/fifo.[ch] could use some improvement, I regret that I will not be able to address that on my current contract.